### ISSUE TITLE
Package iris-vectorlab won't install due to failure to import PyTorch

### ISSUE THREAD

**Issue by isc-shuliu** [intersystems/ipm#546]
## Context
The 3rd party library package is hosted [here](https://github.com/mwaseem75/iris-VectorLab)
See also https://github.com/intersystems/ipm/issues/500#issuecomment-2253097620

## Cause
With some debugging, the bug seems to be caused by attempting to load PyTorch C extension when running `import torch`
```
ImportError: Failed to load PyTorch C extensions:
    It appears that PyTorch has loaded the `torch/_C` folder
    of the PyTorch repository rather than the C extensions which
    are expected in the `torch._C` namespace. This can occur when
    using the `install` workflow. e.g.
        $ python setup.py install && python -c "import torch"
    This error can generally be solved using the `develop` workflow
        $ python setup.py develop && python -c "import torch"  # This should succeed
    or by running Python from a different directory.
```

I get the same error when running `import torch` in interactive `:py` shell. 

## Env
```
IRIS for UNIX (Apple macOS for Apple Silicon) 2024.2.0L (Build 213U) Thu May 30 2024 00:00:30 EDT
```

## Extra
Not sure how it passes on master branch. Possibly because of IRIS version?

**Comment by isc-shuliu** [comment]
Some further debugging suggests that I cannot even install `numpy` in a fresh IRIS instance. Complete error message:
```
>>> import numpy
Traceback (most recent call last):
  File "/opt/ipm09/mgr/python/numpy/_core/__init__.py", line 23, in <module>
    from . import multiarray
  File "/opt/ipm09/mgr/python/numpy/_core/multiarray.py", line 10, in <module>
    from . import overrides
  File "/opt/ipm09/mgr/python/numpy/_core/overrides.py", line 8, in <module>
    from numpy._core._multiarray_umath import (
ModuleNotFoundError: No module named 'numpy._core._multiarray_umath'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/ipm09/mgr/python/numpy/__init__.py", line 114, in <module>
    from numpy.__config__ import show as show_config
  File "/opt/ipm09/mgr/python/numpy/__config__.py", line 4, in <module>
    from numpy._core._multiarray_umath import (
  File "/opt/ipm09/mgr/python/numpy/_core/__init__.py", line 49, in <module>
    raise ImportError(msg)
ImportError:
IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!
Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.
We have compiled some common reasons and troubleshooting tips at:
    https://numpy.org/devdocs/user/troubleshooting-importerror.html
Please note and check the following:
  * The Python version is: Python3.11 from "/opt/ipm09/bin/irisdb"
  * The NumPy version is: "2.1.0"
and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.
Original error was: No module named 'numpy._core._multiarray_umath'
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/opt/ipm09/mgr/python/numpy/__init__.py", line 119, in <module>
    raise ImportError(msg) from e
ImportError: Error importing numpy: you should not try to import numpy from
        its source directory; please exit the numpy source tree, and relaunch
        your python interpreter from there.
```

**Comment by isc-tleavitt** [comment]
@isc-shuliu I'd expect something like this from a python version mismatch between pip install and IRIS.

**Comment by isc-tleavitt** [comment]
Can you run this and paste in output?
```
do ##class(%SYS.Python).GetPythonInfo(.info)
zw info
```

**Comment by isc-tleavitt** [comment]
Also @isc-shuliu I'd be curious to see what the behavior is on 2024.1 given the Flexible Python Runtime changes.

**Comment by isc-shuliu** [comment]
@isc-tleavitt 
I'm still getting the same error on 2024.1. Also according to [IRIS docs](https://docs.intersystems.com/iris20242/csp/docbook/DocBook.UI.Page.cls?KEY=ISP_technologies#ISP_other) we don't support flexible python on macOS yet, even with 2024.2.

## Python Info
On IRIS 2024.2
```
USER>do ##class(%SYS.Python).GetPythonInfo(.info)  zw info
info("AllowNonSystemPythonForIntegratedML")=0
info("CPF_PythonPath")=""
info("CPF_PythonRuntimeLibrary")=""
info("CPF_PythonRuntimeLibraryVersion")=""
info("IRISInsidePython")=0
info("PythonInsideIRIS")=6
info("RunningLibrary")="/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/config-3.11-darwin/libpython3.11.dylib"
info("RunningVersion")="3.11.9 (main, Apr  2 2024, 08:25:04) [Clang 15.0.0 (clang-1500.3.9.4)]"
info("SystemPath")="/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/config-3.11-darwin"
info("SystemVersion")="3.11.6 (main, Oct  2 2023, 20:46:17) [Clang 14.0.0 (clang-1400.0.29.202)]"
info("SystemVersionShort")="3.11.6"
info("iris_site.py_platform")="macos"
```

On IRIS 2024.1
```
USER>do ##class(%SYS.Python).GetPythonInfo(.info)  zw info
info("AllowNonSystemPythonForIntegratedML")=0
info("CPF_PythonPath")=""
info("CPF_PythonRuntimeLibrary")=""
info("RunningLibrary")="/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/config-3.11-darwin/libpython3.11.dylib"
info("RunningVersion")="3.11.9 (main, Apr  2 2024, 08:25:04) [Clang 15.0.0 (clang-1500.3.9.4)]"
info("SystemPath")="/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/config-3.11-darwin"
info("SystemVersion")="3.11.6 (main, Oct  2 2023, 20:46:17) [Clang 14.0.0 (clang-1400.0.29.202)]"
info("SystemVersionShort")="3.11.6"
```

**Comment by isc-shuliu** [comment]
So I inspected the `mgr/python/numpy/core/` folder and found found only Python source code and stub files. 

In contrast, when I installed the file in a python virtual environment (no IRIS or any InterSystems things involved), everything works. In the corresponding directory, I see a couple extra shared-object files compiled from C, including `_multiarray_umath.cpython-312-darwin.so`, which IRIS python couldn't import. (The error message was `ModuleNotFoundError: No module named 'numpy._core._multiarray_umath'`)

**Comment by isc-tleavitt** [comment]
In the GetPythonInfo output, the Clang version mismatch seems pertinent to the error. But I have no idea how to fix this.

**Comment by isc-shuliu** [comment]
After some investigation, I believe I have identified the root cause.

## Cause
There is a Python runtime library that IRIS is linked to, as indicated by the output of `GetPythonInfo()` and the [IRIS documentation](https://docs.intersystems.com/iris20242/csp/docbook/DocBook.UI.Page.cls?KEY=ISP_technologies#ISP_other). It is utilizing Python 3.11 for macOS.

Many Python packages implement their core functionality using [C/C++ extensions](https://docs.python.org/3/extending/extending.html) to achieve fast performance. This process involves compiling the C extension files into shared objects that are both OS-dependent and Python ABI-dependent. Prior to distribution, developers typically compile these sources for each platform and package them as `.whl` wheels (essentially zipped files with a different extension) when uploading to pypi.org. Consequently, for the same version of a package like numpy, there exist various `.whl` distributions, such as `numpy-2.1.0-cp311-cp311-macosx_14_0_arm64.whl` for Python 3.11 and `numpy-2.1.0-cp312-cp312-macosx_14_0_arm64.whl` for Python 3.12.

The issue arises during the installation of these packages using `pip`. The `pip` executable relies on a Python runtime and accesses the python version associated with that runtime. When specifying `--target <IRIS-ROOT>/mgr/python`, it fails to recognize the IRIS python version. Instead, `pip` assumes that the package being installed corresponds to its own python runtime and attempts to install it in a different location. In my scenario, I am utilizing the latest Python 3.12, leading to an attempt to install the numpy wheel specifically compiled for Python 3.12.

## Solution
The manual resolution involves specifying `--python-version` and `--only-binary` during a pip install:
```bash
pip install numpy --python-version 3.11 --only-binary=:all: --target <IRIS-ROOT>/mgr/python
```

Within the context of IPM, we should identify the embedded python version using `sys.version` or `sys.version_info`, and subsequently execute `pip install` with the appropriate `--python-version`.

## Further Discussion
1. For users employing IRIS 2024.2 on platforms supporting BYOP, their installed packages may encounter issues when transitioning between different python versions unless they meticulously manage their sys.path and maintain separate python packages for each python version during `pip install`. This issue impacts the entire IRIS community, irrespective of IPM.
2. I am still puzzled by the absence of shared objects here: https://github.com/intersystems/ipm/issues/546#issuecomment-2297142959. Based on my understanding, I should observe shared objects of the incorrect python version (3.12) in `mgr/python/numpy/_core/`. Upon retrying with `pip install numpy --no-binary=numpy --target <IRIS-ROOT>/mgr/python`, I indeed see the Python 3.12 objects as anticipated. There appears to be a gap in my comprehension, although it does not appear to be a pressing concern at the moment.

**Comment by isc-shuliu** [comment]
@isc-tleavitt 

> Within the context of IPM, we should identify the embedded python version using sys.version or sys.version_info, and subsequently execute pip install with the appropriate --python-version.

This also introduces the headache of accessing the current python version. On IRIS instances with `##class(%SYS.Python).GetPythonInfo()`, we can access this quite easily.

Otherwise, we need to call `python -c "import sys; print(sys.version)"` or `python --version` using `$zf(-100)`. But the exact name and location of `python` still needs to be resolved based either on user input or $PATH. Are there any smarter ways to do this?

**Comment by isc-shuliu** [comment]
Never mind, figured out a way
```objectscript
set sysmodule = ##class(%SYS.Python).Import("sys")
w sysmodule."version_info".major
```

**Comment by isc-shuliu** [comment]
This is being fixed by #538

**Comment by isc-shuliu** [comment]
Fixed in #538 and #543
